### DISCUSSION TITLE
How can I list unused macros?

### DISCUSSION THREAD

**Discussion by oharboe** [The-OpenROAD-Project/OpenROAD#5009] [General]
I would like to check if I have some unused macros in the ADDITIONAL_* macros.

How can I do that?

This is for [megaboom](https://github.com/The-OpenROAD-Project/megaboom) that has a lot of SRAMs and I want to make sure that I don't provide SRAMs that are not used.

**Comment by rovinski** [comment]
This would be easier in Tcl if `dbBlock.getMasters()` returned a swig-convertible type rather than requiring a `vector<dbMaster*>` to be passed, because I don't know how to create that in Tcl (@maliberty?). For that solution, you would simply get the masters from the libraries and then filter out the masters from the block, because the dbBlock.getMasters() only returns masters used by the block.

The other (less efficient) approach is to iterate through the instances and filter out their masters one-by-one, leaving only the unused ones left:
```
set db [ord::get_db]
set libs [$db getLibs]
set masters ""
foreach lib $libs {
  foreach master [$lib getMasters] {
    # filter out non-block masters, or you can remove this conditional to detect any unused master
    if {[$master getType] == "BLOCK"} {
      lappend masters $master
    }
  }
}

set block [ord::get_db_block]
set insts [$block getInsts]

foreach inst $insts {
  set inst_master [$inst getMaster]
  set masters [lsearch -all -not -inline $masters $inst_master]
}

foreach master $masters {
  puts "Master [$master getName] is loaded but not used in the design"
}
```

**Reply by oharboe** [reply]
Worked: https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/pull/1979
